File: python/grid.py
from cell import Cell


class Grid:
    rows = 0
    columns = 0

    def __init__(self, rows, columns):
        self.rows = rows
        self.columns = columns

        self.grid = self.prepare_grid()
        self.configure_cells()

    def prepare_grid(self):
        return [[Cell(j, i) for i in range(self.columns)] for j in range(self.rows)]

    def configure_cells(self):
        for row in range(self.rows):
            for col in range(self.columns):
                cell = self.grid[row][col]
                cell.north = self[row - 1, col]
                cell.south = self[row + 1, col]
                cell.east = self[row, col + 1]
                cell.west = self[row, col - 1]

    def __getitem__(self, item):
        if isinstance(item, tuple) and len(item) == 2:
            row, col = item

            if 0 <= row < self.rows and 0 <= col < self.columns:
                return self.grid[row][col]
            else:
                return None

    def each_cell(self):
        for row in range(self.rows):
            for col in range(self.columns):
                yield self[row, col]

    def __str__(self):
        output = "+" + "---+" * self.columns + "\n"
        for row in self.grid:
            top = "|"
            bottom = "+"
            for cell in row:
                if cell is None:
                    cell = Cell(-1, 1)
                body = "   "
                east_boundary = " " if cell.linked(cell.east) else "|"
                top += body + east_boundary
                south_boundary = "   " if cell.linked(cell.south) else "---"
                bottom += south_boundary + "+"
            output += top + "\n" + bottom + "\n"

        return output

File: go/maze/grid.go
package maze

import (
	"strings"
)

type Grid struct {
	rows    int
	columns int
	grid    [][]*Cell
}

func NewGrid(rows int, columns int) *Grid {
	grid := Grid{rows: rows, columns: columns}
	grid.Prepare()
	grid.Configure()
	return &grid
}

func (grid *Grid) Prepare() {
	grid.grid = make([][]*Cell, grid.rows)
	for row := range grid.rows {
		grid.grid[row] = make([]*Cell, grid.columns)
		for col := range grid.columns {
			cell := newCell(row, col)
			grid.grid[row][col] = cell
		}
	}
}

func (grid *Grid) Configure() {
	for row := range grid.rows {
		for col := range grid.columns {
			cell := grid.GetCell(row, col)
			cell.north = grid.GetCell(row-1, col)
			cell.south = grid.GetCell(row+1, col)
			cell.east = grid.GetCell(row, col+1)
			cell.west = grid.GetCell(row, col-1)
		}
	}
}

func (grid *Grid) GetCell(row int, col int) *Cell {
	if row < 0 || row >= grid.rows || col < 0 || col >= grid.columns {
		return nil
	}

	return grid.grid[row][col]
}

func (grid *Grid) String() string {
	output := "+" + strings.Repeat("---+", grid.columns) + "\n"
	for row := range grid.rows {
		top := "|"
		bottom := "+"
		for _, cell := range grid.grid[row] {
			if cell == nil {
				cell = newCell(-1, -1)
			}
			body := "   "
			easternBoundary := "|"

			if cell.linked(cell.east) {
				easternBoundary = " "
			}
			top += body + easternBoundary
			southBoundary := "---"

			if cell.linked(cell.south) {
				southBoundary = "   "
			}
			bottom += southBoundary + "+"
		}
		output += top + "\n" + bottom + "\n"
	}

	return output
}
